src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,432 @@
+
+    /** @param httpStack HTTP stack to be used */
+    public BasicAsyncNetwork(BaseHttpStack httpStack) {
+        // If a pool isn't passed in, then build a small default pool that will give us a lot of
+        // benefit and not use too much memory.
+        this(httpStack, new ByteArrayPool(DEFAULT_POOL_SIZE));
+    }
+
+    /**
+     * @param httpStack HTTP stack to be used
+     * @param pool a buffer pool that improves GC performance in copy operations
+     */
+    public BasicAsyncNetwork(BaseHttpStack httpStack, ByteArrayPool pool) {
+        mBaseHttpStack = httpStack;
+        // Populate mHttpStack for backwards compatibility, since it is a protected field. However,
+        // we won't use it directly here, so clients which don't access it directly won't need to
+        // depend on Apache HTTP.
+        mHttpStack = httpStack;
+        mPool = pool;
+        mHandler = new Handler(Looper.myLooper());
+    }
+
+    /* Method to be called after a successful network request */
+    private void onRequestSucceeded(
+            final Request<?> request,
+            final long requestStartMs,
+            final HttpResponse httpResponse,
+            final OnRequestComplete callback) {
+        final int statusCode = httpResponse.getStatusCode();
+        final List<Header> responseHeaders = httpResponse.getHeaders();
+        // Handle cache validation.
+        if (statusCode == HttpURLConnection.HTTP_NOT_MODIFIED) {
+            Entry entry = request.getCacheEntry();
+            if (entry == null) {
+                callback.onSuccess(
+                        new NetworkResponse(
+                                HttpURLConnection.HTTP_NOT_MODIFIED,
+                                /* data= */ null,
+                                /* notModified= */ true,
+                                SystemClock.elapsedRealtime() - requestStartMs,
+                                responseHeaders));
+                return;
+            }
+            // Combine cached and response headers so the response will be complete.
+            List<Header> combinedHeaders = NetworkUtility.combineHeaders(responseHeaders, entry);
+            callback.onSuccess(
+                    new NetworkResponse(
+                            HttpURLConnection.HTTP_NOT_MODIFIED,
+                            entry.data,
+                            /* notModified= */ true,
+                            SystemClock.elapsedRealtime() - requestStartMs,
+                            combinedHeaders));
+            return;
+        }
+
+        byte[] responseContents = httpResponse.getContentBytes();
+
+        if (responseContents == null) {
+            final InputStream inputStream = httpResponse.getContent();
+            if (inputStream == null) {
+                // Add 0 byte response as a way of honestly representing a
+                // no-content request.
+                responseContents = new byte[0];
+            } else {
+                Runnable run =
```

Comment:
A lot of this is copied from the BasicNetwork stream copying code, right? Can we share it?